source_code/main_mcu/src/fido2/fido2.c

```diff
@@ -0,0 +1,493 @@
+    ecc256_pub_key pub_key;
+    uint32_t confirmation;
+
+    memset(&auth_data_header, 0, sizeof(auth_data_header));
+
+    if (rec == NULL)
+    {
+        response->error_code = CRED_NOT_FOUND;
+        return response->error_code;
+    }
+    else
+    {
+    }
+    crypto_sha256_init();
+    crypto_sha256_update(request->rpID, strnlen(request->rpID, RPID_LEN));
+    crypto_sha256_final(auth_data_header.rpID_hash);
+
+    confirmation = fido2_prompt_user("Confirm auth?", TRUE, 10);
+
+    if (confirmation == SUCCESS)
+    {
+        auth_data_header.flags |= UP_BIT; //User present
+        auth_data_header.flags |= UV_BIT; //User verified
+        if (request->new_cred)
+        {
+            auth_data_header.flags |= AT_BIT; //Attested credential data attached
+        }
+        response->error_code = SUCCESS;
+    }
+    else
+    {
+        auth_data_header.flags &= ~UP_BIT;
+        auth_data_header.flags &= ~UV_BIT;
+        auth_data_header.flags &= ~AT_BIT;
+        response->error_code = (uint8_t) OPERATION_DENIED;
+        return response->error_code;
+    }
+
+    //Create public key from stored private key
+    crypto_ecc256_derive_public_key(rec->priv_key, &pub_key);
+
+    //Update credential "count"
+    //Update on every get_assertion/get_next_assertion
+    rec->count = fido2_update_credential_count(rec->count);
+
+    fido2_update_credential_in_db(rec);
+
+    //Create attestation signature
+    fido2_set_attest_sign_count(rec->count, &auth_data_header.sign_count);
+
+    crypto_ecc256_load_key(rec->priv_key);
+    fido2_calc_attestation_signature((uint8_t const *) &auth_data_header, sizeof(auth_data_header), request->client_data_hash, response->attest_sig);
+
+    //Fill out response object
+    memcpy(response->tag, rec->tag, TAG_LEN);
+    memcpy(response->rpID_hash, auth_data_header.rpID_hash, RPID_HASH_LEN);
+    response->count_BE = auth_data_header.sign_count;
+    memcpy(response->pub_key_x, pub_key.x, PUB_KEY_X_LEN);
+    memcpy(response->pub_key_y, pub_key.y, PUB_KEY_Y_LEN);
+    //Attest signature already filled out above when calculating attestation signature
+    memcpy(response->aaguid, MINIBLE_AAGUID, AAGUID_LEN);
+    //We no longer have the credential ID. Not needed for assertion
+    response->cred_ID_len = 0;
+    response->flags = auth_data_header.flags;
+    //error code already set
+    return response->error_code;
+}
+
+/*! \fn     fido2_process_make_auth_data(FIDO2_make_auth_data_req_message_t const *request, FIDO2_make_auth_data_rsp_message_t *response)
+*   \brief  Make the authentication data. This esentially creates the key pair and stores
+*           the new record in the DB
+*   \param  incoming request messsage
+*   \param  outgoing response message
+*   \return void
+*/
+void fido2_process_make_auth_data(FIDO2_make_auth_data_req_message_t const *request, FIDO2_make_auth_data_rsp_message_t *response)
+{
+    memset(response, 0, sizeof(*response));
+
+    if (request->new_cred == 1)
```

Comment:
really not a big fan of magic numbers... maybe a comment?

Reply:
Replaced with #define